src/SnippetProvider/RaygunSnippetProvider.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+namespace SilverStripe\TagManager\SnippetProvider;
+
+use SilverStripe\Core\Environment;
+use SilverStripe\TagManager\SnippetProvider;
+use SilverStripe\Forms\FieldList;
+use SilverStripe\Forms;
+use SilverStripe\Core\Manifest\ModuleResourceLoader;
+
+/**
+ * A snippet provider that lets you add Raygun
+ */
+class RaygunSnippetProvider implements SnippetProvider
+{
+    public function getTitle()
+    {
+        return "Raygun crash reporting (JavaScript front-end)";
+    }
+
+    public function getParamFields()
+    {
+        $apiKeyDefined = !empty(Environment::getEnv('SS_RAYGUN_APP_KEY'));
+
+        $imgSrc = ModuleResourceLoader::singleton()->resolveURL('sminnee/tagmanager:client/img/raygun-guide.png');
+
+        $fields = [];
+        if ($apiKeyDefined) {
+            $fields[] = (new Forms\CheckboxField('ENV_APP_KEY', 'Use "silverstripe/raygun" module'))
+                ->setDescription('Choose to reuse "silverstripe/raygun" module settings. This is the default and you want this in most cases');
+        }
+
+        if ($apiKeyDefined) {
+            $apiKeyTitle = 'API Key (only if you didn\'t select the checkbox above)';
```

Comment:
Seems like a good use case for [display-logic](http://github.com/unclcheese/silverstripe-display-logic).

```php
$field->displayIf('ENV_APP_KEY')->isChecked()
```

It adds a dependency, but it has a very wide install base anyway.

Reply:
ok, I've had a tough fight with the `ParamExpander` trait, and after about 4 hours of trying to make it all work together I think we should refactor the module and remove the trait altogether, because it gets in a way and doesn't let to work with the form fields in a predictable way, it doesn't handle CompositeFields properly and it obscures the API of the module too much.

However, for this particular PR I think it's easier to leave all the fields on the page as is, otherwise it converts from a mere snapshot provider implementation into lots of refactoring.

Reply:
Just discussed in person that custom javascript is also not worth it, because we cannot use CompositeField (ParamExpander loses those field values and we end up without saving them to the database)